esp-hal/src/lib.rs

```diff
@@ -572,6 +572,115 @@ mod critical_section_impl {
     }
 }
 
+// The state of a re-entrant lock
+pub(crate) struct LockState {
+    #[cfg(multi_core)]
+    core: portable_atomic::AtomicU8,
+}
+
+impl LockState {
+    pub const fn new() -> Self {
+        Self {
+            #[cfg(multi_core)]
+            core: portable_atomic::AtomicU8::new(0),
+        }
+    }
+
+    #[cfg(multi_core)]
+    fn current_core() -> u8 {
+        // Note: 0 means unlocked.
+        match get_core() {
```

Comment:
I'd suggest not spending CPU time mapping one core indexing scheme to the other.

Reply:
Updated